### PR TITLE
[VP][RISCV] Introduce vp.splat and RISC-V.

### PR DESCRIPTION
**Pull request by yetingk**
This patch introduces a vp intrinsic for splat. It's helpful for IR-level passes to create a splat with specific vector length.

### CODE REVIEW

File: llvm/docs/LangRef.rst [hunk 1]
```diff
@@ -22841,6 +22841,51 @@ Examples:
...
+'``llvm.experimental.vp.splat``' Intrinsic
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Syntax:
+"""""""
+This is an overloaded intrinsic.
+
+::
+
+      declare <2 x double> @llvm.experimental.vp.splat.v2f64(<2 x double> %vec, <2 x i1> %mask, i32 %evl)
```
Comment (frasercrmck) [inline llvm/docs/LangRef.rst:22856]: This should probably be something like `(double %scalar, <2 x i1> %mask, i32 %evl)`

File: llvm/include/llvm/IR/VPIntrinsics.def [hunk 2]
```diff
@@ -777,6 +777,13 @@ END_REGISTER_VP(experimental_vp_reverse, EXPERIMENTAL_VP_REVERSE)
 
 ///// } Shuffles
 
+// llvm.vp.splat(ptr,val,mask,vlen)
```
Comment (frasercrmck) [inline llvm/include/llvm/IR/VPIntrinsics.def:780]: `llvm.experimental.vp.splat(val,mask,vlen)` or `x` or something.

**Review by frasercrmck — COMMENTED** [review]
It would be quite nice to see how this is expected to help codegen where we currently use regular spats. I assume it will likely reduce `vsetvli`s? Is there a plan to introduce it to the vectorizer?

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -22841,6 +22841,51 @@ Examples:
...
+      declare <vscale x 4 x i32> @llvm.experimental.vp.splat.nxv4i32(<vscale x 4 x i32> %vec, <vscale x 4 x i1> %mask, i32 %evl)
+
+Overview:
+"""""""""
+
+The '``llvm.experimental.vp.splat.*``' intrinsic is to create a prdicated splat
+with specific effective vector length.
```
Comment (victor-eds) [inline llvm/docs/LangRef.rst:22863]:
```suggestion
The '``llvm.experimental.vp.splat.*``' intrinsic is to create a predicated splat
with specific effective vector length.
```

File: llvm/docs/LangRef.rst [hunk 4]
```diff
@@ -22841,6 +22841,51 @@ Examples:
...
+
+Arguments:
+""""""""""
+
+The result is a vector and it is a splat of the second scalar operand. The
```
Comment (topperc) [inline llvm/docs/LangRef.rst:22868]: operand -> argument

Comment (topperc) [reply llvm/docs/LangRef.rst:22868]: It's the first operand not the second right?

File: llvm/docs/LangRef.rst [hunk 5]
```diff
@@ -22841,6 +22841,51 @@ Examples:
...
+second argument ``mask`` is a vector mask and has the same number of elements as
+the result. The third argument is the explicit vector length of the operation.
+
+Semantics:
+""""""""""
+
+This intrinsic splats a vector with ``evl`` elements of a scalar operand.
```
Comment (topperc) [inline llvm/docs/LangRef.rst:22875]: operand -> argument

File: llvm/docs/LangRef.rst [hunk 6]
```diff
@@ -22841,6 +22841,51 @@ Examples:
...
+elements past ``evl`` are poison.
+
+Examples:
+"""""""""
+
+.. code-block:: llvm
+
+      %r = call <4 x float> @llvm.vp.splat.v4f32(float %a, <4 x i1> %mask, i32 %evl)
+      ;; For all lanes below %evl, %r is lane-wise equivalent to %also.r
+      %also.r = select <4 x i1> %mask, <4 x float> splat(float %a), <4 x float> poison
```
Comment (topperc) [inline llvm/docs/LangRef.rst:22886]: The `splat` syntax is only valid for constants. Should this use the insertelt+shufflevector syntax?

File: llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp [hunk 7]
```diff
@@ -2211,10 +2213,14 @@ SDValue DAGTypeLegalizer::PromoteIntOp_INSERT_VECTOR_ELT(SDNode *N,
 }
 
 SDValue DAGTypeLegalizer::PromoteIntOp_ScalarOp(SDNode *N) {
+  SDValue Op = GetPromotedInteger(N->getOperand(0));
+  if (N->getOpcode() == ISD::EXPERIMENTAL_VP_SPLAT)
+    return DAG.getNode(ISD::EXPERIMENTAL_VP_SPLAT, SDLoc(N), N->getValueType(0),
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:2218]: Can we use UpdateNodeOperands?

File: llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp [hunk 8]
```diff
@@ -5859,7 +5866,11 @@ SDValue DAGTypeLegalizer::PromoteIntRes_ScalarOp(SDNode *N) {
   EVT NOutElemVT = NOutVT.getVectorElementType();
 
   SDValue Op = DAG.getNode(ISD::ANY_EXTEND, dl, NOutElemVT, N->getOperand(0));
-
+  if (N->isVPOpcode()) {
+    SDValue Mask = N->getOperand(1);
```
Comment (topperc) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:5870]: Can we pass N->getOperand(1) and N->getOperand(2) directly to getNode without the temporaries?

**Comment by lukel97** [comment]
> It would be quite nice to see how this is expected to help codegen where we currently use regular spats. I assume it will likely reduce `vsetvli`s? Is there a plan to introduce it to the vectorizer?

In https://github.com/llvm/llvm-project/pull/98140#discussion_r1670337004 we ran into an awkward spot where we wanted to perform a splat with an AVL to avoid vl toggles. I think the canonical form would still be shufflevector (or splat_vector for scalable vectors at the SelectionDAG level)

**Review by topperc — APPROVED** [review]
LGTM

**Review by michaelmaitland — APPROVED** [review]
LGTM with minor comment

File: llvm/docs/LangRef.rst [hunk 9]
```diff
@@ -22841,6 +22841,53 @@ Examples:
...
+
+The '``llvm.experimental.vp.splat.*``' intrinsic is to create a predicated splat
+with specific effective vector length.
+
+Arguments:
+""""""""""
+
+The result is a vector and it is a splat of the first scalar argument. The
+second argument ``mask`` is a vector mask and has the same number of elements as
+the result. The third argument is the explicit vector length of the operation.
```
[the remark this answers is not in the document]
Comment (topperc) [reply llvm/docs/LangRef.rst:22870]: Yes. It's documented for all VP intrinsics at the start of the section https://llvm.org/docs/LangRef.html#vector-predication-intrinsics

**Pull request merged**